src/Symfony/Component/Cache/Adapter/TagAwareAdapter.php

```diff
@@ -406,97 +312,95 @@ public function __destruct()
-            foreach ($items as $key => $item) {
-                foreach ($itemTagVersions[$key] as $itemTag => $itemTagVersion) {
-                    if (!isset($tagVersions[$itemTag]) || $tagVersions[$itemTag] !== $itemTagVersion) {
-                        (self::$unsetCacheItem)($item);
-                        break;
-                    }
+
+        $tagVersions = $this->getTagVersions($itemTags);
+        foreach ($itemTags as $key => $tags) {
+            foreach ($tags as $tag => $version) {
+                if ($tagVersions[$tag] !== $version) {
+                    unset($itemTags[$key]);
+                    continue 2;
                 }
-                yield $key => $item;
-            }
-        } else {
-            foreach ($items as $key => $item) {
-                yield $key => $item;
             }
         }
+        $tagVersions = null;
+
+        yield from (self::$yieldItems)($bufferedItems, $itemTags);
     }
 
-    /**
-     * Loads tag versions from or creates them in the tag pool, and updates the cache of known tag versions.
-     *
-     * May return only a part of requested tags or even none of them if for some reason they cannot be read or created.
-     *
-     * @throws InvalidArgumentException
-     *
-     * @return string[]
-     */
-    private function getTagVersions(array $tags): array
+    private function getTagVersions(array $tagsByKey): array
     {
-        if (!$tags) {
-            return [];
+        $tagVersions = [];
+        $fetchTagVersions = false;
+
+        foreach ($tagsByKey as $tags) {
+            $tagVersions += $tags;
+
+            foreach ($tags as $tag => $version) {
+                if ($tagVersions[$tag] !== $version) {
+                    unset($this->knownTagVersions[$tag]);
+                }
+            }
         }
 
-        $tagIdsMap = $tagVersions = $createdTagVersions = $createdTags = [];
-        foreach ($tags as $tag) {
-            $tagIdsMap[static::TAG_PREFIX.$tag] = $tag;
+        if (!$tagVersions) {
+            return [];
         }
-        ksort($tagIdsMap);
 
-        if (0.0 < $this->knownTagVersionsTtl) {
-            $now = microtime(true);
-            $knownTagVersionsExpiration = $now + $this->knownTagVersionsTtl;
-            if (self::MAX_NUMBER_OF_KNOWN_TAG_VERSIONS < \count($this->knownTagVersions)) {
-                $this->knownTagVersions = array_filter($this->knownTagVersions, static function ($v) use ($now) { return $now < $v[0]; });
-            }
-            foreach ($this->tags->getItems(array_keys($tagIdsMap)) as $tagId => $version) {
-                $tag = $tagIdsMap[$tagId];
-                if ($version->isHit()) {
-                    $tagVersions[$tag] = $version->get();
-                    $this->knownTagVersions[$tag] = [$knownTagVersionsExpiration, $tagVersions[$tag]];
-                    continue;
-                }
-                $createdTags[] = $version->set($newTagVersion ??= random_bytes(8));
```

Comment:
I prefer keeping `random_int` for that. The reason is that igbinary can pack them more efficiently when one cares about about payload size.

Reply:
On 64-bit platforms, only 2^32 of 2^64 (i.e. 1 of 2^32) can be packed into smaller size. For strings, igbinary has `compact_strings` option which is On by default. But this works only when more than one tag is generated and attached to an item. It was not obvious which type to choose, so the last argument for me was that on 32-bit platforms random_int() is only 4 bytes which is 1. weak and 2. cannot be mixed with 64-bit platforms.

Reply:
Btw I've never seen 32-bit php since win98 :)

Reply:
OK, I propose `random_bytes(6)` then
(some ppl run php on raspberry 32b :) )

Reply:
updated